### ISSUE TITLE
Management of dynamic PMT

### ISSUE THREAD

**Issue by thibautperrin** [google/ExoPlayer#9109]
Hello,
I am building a DVB-T TV player, with Exoplayer.
I recently discovered how some TV channels, like Canal+, proceed to be in clear mode at some time and in crypted mode at other time.
Actually, the stream PIDs in the PMT are changing, (but the PMT's PID remains the same).
The problem is that my DVB scan is static, and when I try to watch this channel, I would need to retrieve the PMT content from Exoplayer, to adjust the stream PID that I filter from my tuner.
Is there a way to retrieve the stream PIDs from the PMT with Exoplayer ?

Best regards,

Thibaut

**Comment by kim-vde** [comment]
There is currently no way of getting notified of a PMT change in ExoPlayer. Actually, ExoPlayer ignores the PMTs of a given PID after it has parsed the first one.

We can try to fix this but it might take some time before I have time to look at it. Another solution is for you to create your own extractor by forking the one of ExoPlayer (`TsExtractor`) and adapting it to detect the PMT changes. A simpler (but slightly more hacky) solution is to detect when the player is in buffering state for too long.

What do you plan to do once you get notified of a PMT change? You could re-prepare the player, but the transition won’t be seamless.

How often are these PMT changes occuring?

**Comment by thibautperrin** [comment]
Hello,
Thank you for your answer.

> We can try to fix this but it might take some time before I have time to look at it. Another solution is for you to create your own extractor by forking the one of ExoPlayer (`TsExtractor`) and adapting it to detect the PMT changes. A simpler (but slightly more hacky) solution is to detect when the player is in buffering state for too long.
>
I tried using my own TsExtractor, but I had the feeling that TsExtractor was only for TS data, and the PMT is not TS but PSI data. Are you sure that I can parse the PMT by extending the TsExtractor ? I could try again.


> What do you plan to do once you get notified of a PMT change? You could re-prepare the player, but the transition won’t be seamless.
> 
When I will be notified that the PMT has change, I will be able to ask to the tuner to filter on a different set of PIDs based on the new PMT.


 
> How often are these PMT changes occuring?
>
For Canal+, The PMT changes 4-6 times every day.

**Comment by kim-vde** [comment]
> I tried using my own TsExtractor, but I had the feeling that TsExtractor was only for TS data, and the PMT is not TS but PSI data. Are you sure that I can parse the PMT by extending the TsExtractor ? I could try again.

Yes, the `TsExtractor` contains a [`PmtReader`](https://github.com/google/ExoPlayer/blob/r2.14.2/library/extractor/src/main/java/com/google/android/exoplayer2/extractor/ts/TsExtractor.java#L537) class that parses the metadata in this table.

**Comment by thibautperrin** [comment]
> Yes, the `TsExtractor` contains a `PmtReader` class that parses the metadata in this table.
>
Thank you !
I tested by copy-pasting TsExtractor into my code to adapt it and modify the PmtReader and it seems to work : When I detect a video stop, I restart it, and when the video start, the PMT is read, and if the PMT is not as expected, I change the PID list sent to the tuner.

However, it means that I have to copy-paste the totality of 3 exoplayers files, because `TsExtractor` is a final class, and it accesses `TsDurationExtractor` and `TsBinarySearchSeeker` which are package-private. Is there a cleaner way to customize only the `PmtReader` and `PatReader` ?
 
Other point: I can also be notified when the PMT change by executing this code at a regular interval (10 seconds):
`            tsPayloadReaders.put(TS_PAT_PID, new SectionReader(new PatReader()));
`

**Comment by kim-vde** [comment]
> Is there a cleaner way to customize only the PmtReader and PatReader ?

I am afraid not.

> Other point: I can also be notified when the PMT change by executing this code at a regular interval (10 seconds):
tsPayloadReaders.put(TS_PAT_PID, new SectionReader(new PatReader()));

Instead of doing that, you can probably just keep the PatReader in the tsPayloadReaders map instead of removing it when the PAT has been parsed.
